README.md
<iframe src="https://royw.github.io/Versio/" style="width:100%; height:400px;"></iframe>


### Latin

    Noun
    1. a turning, change, version

### About Versio

Versio is a generic version class that supports comparison and version bumping (incrementing by part, for example
you may bump the minor part of '1.2.3' yielding '1.3.0').

Four version schemes are included:

* **Simple3VersionScheme** which supports 3 numerical part versions (A.B.C where A, B, and C are integers)
* **Simple4VersionScheme** which supports 4 numerical part versions (A.B.C.D where A, B, C, and D are integers)
* **Pep440VersionScheme** which supports PEP 440 (http://www.python.org/dev/peps/pep-0440/) versions
  (N[.N]+[{a|b|c|rc}N][.postN][.devN][+local])
* **PerlVersionScheme** which supports 2 numerical part versions where the second part is at least two digits
  (A.BB where A and B are integers and B is zero padded on the left.  For example:  1.02, 1.34, 1.567)

If you don't specify which version scheme the version instance uses, then it will use the first scheme from the
**SupportedVersionSchemes** list that successfully parses the version string

By default, **Pep440VersionScheme** is the supported scheme.  To change to a different list of schemes, use the
**Version.set_supported_version_schemes(schemes)**.  For example::

    from versio.version_scheme import Simple3VersionScheme, PerlVersionScheme
    Version.set_supported_version_schemes([Simple3VersionScheme, PerlVersionScheme])

In addition, you may define your own version scheme by creating a new VersionScheme instance.

The VersionScheme class defines the version scheme.

A version scheme consists of:

* a name,
* a regular expression used to parse the version string,
* the regular expression flags to use (mainly to allow verbose regexes),
* a format string used to reassemble the parsed version into a string,
* an optional list of field types (if not specified, assumes all fields are strings),
* a list of field names used for accessing the components of the version.
* an optional subfield dictionary with the key being a field name, and the value being a list of sub-field names.
  For example, in the **Pep440VersionScheme**, the "Release" field may contain multiple parts, so we use
  subfield names for the parts.  Say we have a version of "1.2.3rc1", the "1.2.3" is the release field, then
  "1" is the "major" subfield, "2" is the "minor" subfield, and "3" is the "tiny" subfield.
* a "clear" value used to set the field values to the right of the field being bumped,
* a sequence dictionary where the keys are the field names, and the values are a list of allowed values.
  The list must be in bump order.  Bumping the last value in the list has no effect.

Note, you need to manually maintain consistency between the regular expression,
the format string, the optional field types, and the fields list.  For example,
if your version scheme has N parts, then the regular expression should match
into N groups, the format string should expect N arguments to the str.format()
method, and there must be N unique names in the fields list.

### Installation

From pypi::

    pip install Versio


### Usage

First let's just play with the comparing and bumping versions::

    >>> from versio.version import Version
    >>> from versio.version_scheme import Pep440VersionScheme
    >>> v1 = Version('1.2.3rc4.post5.dev6', scheme=Pep440VersionScheme)
    >>> v2 = Version('1.2.3rc4.post5.dev7', scheme=Pep440VersionScheme)
    >>> v1 == v2
    False
    >>> v1 < v2
    True
    >>> v1.bump('dev')
    True
    >>> str(v1)
    '1.2.3rc4.post5.dev7'
    >>> v1 == v2
    True
    >>> v1.bump('pre')
    True

Now let's look in the PEP 440 scheme::

    >>> Pep440VersionScheme.fields
    ['release', 'pre', 'post', 'dev', 'local']
    >>> Pep440VersionScheme.format_str
    '{0}{1}{2}{3}{4}'

The fields are used by the **bump(field)** method in the above example.  We skipped bumping the release above so let's
do it here::

    >>> v1 = Version('1.2.3rc4.post5.dev6', scheme=Pep440VersionScheme)
    >>> str(v1)
    '1.2.3rc4.post5.dev6'
    >>> v1.bump('release')
    True
    >>> str(v1)
    '1.2.4'

For PEP 440, the release part is defined as "N[.N]+".  We can bump specific parts of the release by using an
index like::

    >>> v1.bump('release', 2)
    True
    >>> str(v1)
    '1.2.5'
    >>> v1.bump('release', 1)
    True
    >>> str(v1)
    '1.3.0'
    >>> v1.bump('release', 0)
    True
    >>> str(v1)
    '2.0.0'

To use a name directly, we use the concept of subfields which are mapped to a field/index pair::

    >>> Pep440VersionScheme.subfields
    {'tiny2': ['Release', 3], 'major': ['Release', 0], 'tiny': ['Release', 2], 'minor': ['Release', 1]}

    >>> v1 = Version('1.2.3rc4.post5.dev6', scheme=Pep440VersionScheme)
    >>> str(v1)
    '1.2.3rc4.post5.dev6'
    >>> v1.bump('tiny')
    True
    >>> str(v1)
    '1.2.4'
    >>> v1.bump('minor')
    True
    >>> str(v1)
    '1.3.0'
    >>> v1.bump('major')
    True
    >>> str(v1)
    '2.0.0'

Now that you've seen the version scheme in action, let's take a look at how it is defined::

    Pep440VersionScheme = VersionScheme(name="pep440",
                                        parse_regex=r"""
                                        ^
                                        (\d[\.\d]*(?<= \d))
                                        ((?:[abc]|rc)\d+)?
                                        (?:(\.post\d+))?
                                        (?:(\.dev\d+))?
                                        (?:(\+(?![.])[a-zA-Z0-9\.]*[a-zA-Z0-9]))?
                                        $
                                        """,
                                        compare_order=[0, 1, 2, 3, 4],
                                        compare_fill=['~', '~', '', '~', ''],
                                        parse_flags=re.VERBOSE,
                                        clear_value=None,
                                        format_str='{0}{1}{2}{3}{4}',
                                        fields=['Release', 'Pre', 'Post', 'Dev', 'Local'],
                                        subfields={'Release': ['Major', 'Minor', 'Tiny', 'Tiny2']},
                                        sequences={'Pre': ['a', 'b', 'c', 'rc'],
                                                   'Post': ['.post'],
                                                   'Dev': ['.dev'],
                                                   'Local': ['+']},
                                        description=dedent("""\
                                            PEP 440
                                            Public version identifiers MUST comply with the following scheme:

                                            N[.N]+[{a|b|c|rc}N][.postN][.devN][+local]

                                            Public version identifiers MUST NOT include leading or trailing whitespace.

                                            Public version identifiers MUST be unique within a given distribution.

                                            Public version identifiers are separated into up to five segments:

                                                Release segment: N[.N]+
                                                Pre-release segment: {a|b|c|rc}N
                                                Post-release segment: .postN
                                                Development release segment: .devN
                                                Local release segment: +local

                                            The local version labels MUST be limited to the following set of permitted
                                            characters:

                                                ASCII letters ( [a-zA-Z] )
                                                ASCII digits ( [0-9] )
                                                periods ( . )

                                            Local version labels MUST start and end with an ASCII letter or digit.
                                        """))

The **parse_regex** and **parse_flags** do what you think by parsing a string into a list containing regex groups,
except the group list is zero indexed to the first matching group.

The **format_str** and **format_types** control how the version is converted to a string in **__str__()**.  Basically
**format_str.format(*args)** is called where the args is a list built by casting each of the version's groups using
the corresponding type from the  **format_types** list.  If you don't specify a **format_types**, then each group
is cast as a str.

The **clear_value** typically should be '0' for numeric versions and None for non-numeric.  Basically it specifies
what to put in the groups to the right of the group being bumped.

The **sequences** dictionary maps text to prepend to a group when formatting.  The dictionary keys must be in the
**fields** list.  To progress through the sequence, bump the field with an index of 0.  An index of 1 bumps the numeric
part of the group.  For example::

    >>> v1 = Version('1.2.3a4.post5.dev6', scheme=Pep440VersionScheme)
    >>> str(v1)
    '1.2.3a4.post5.dev6'
    >>> v1.bump('pre', 0)
    True
    >>> str(v1)
    '1.2.3b1'
    >>> v1.bump('pre', 1)
    True
    >>> str(v1)
    '1.2.3b2'
    >>> v1.bump('pre', 0)
    True
    >>> str(v1)
    '1.2.3c1'
    >>> v1.bump('pre', 0)
    True
    >>> str(v1)
    '1.2.3rc1'
    >>> v1.bump('pre', 0)
    False
    >>> str(v1)
    '1.2.3rc1'

Notice that bumping fails at the end of the sequence, and the version is not changed.  You can override this behavior
by setting the "promote" argument to True, resulting in the field being removed at the end of the field's sequence.
For example:


    >>> v1 = Version('1.2.3rc1', scheme=Pep440VersionScheme)
    >>> str(v1)
    '1.2.3rc1'
    >>> v1.bump('pre', 0)
    False
    >>> str(v1)
    '1.2.3rc1'
    >>> v1.bump('pre', 0, promote=True)
    True
    >>> str(v1)
    '1.2.3'

That's it.

There are more examples in *tests/version_test.py*.  You may test directly with *py.test* or against multiple
python versions with *tox*.

Enjoy!


news.txt
News
====

Initial Release


versio/version.py
# coding=utf-8
"""
A generic version class that supports comparison and version bumping (incrementing by part, for example you may
bump the minor part of '1.2.3' yielding '1.3.0').

Four version schemes are included::

    * Simple3VersionScheme which supports 3 numerical part versions (A.B.C where A, B, and C are integers)
    * Simple4VersionScheme which supports 4 numerical part versions (A.B.C.D where A, B, C, and D are integers)
    * Pep440VersionScheme which supports PEP 440 (http://www.python.org/dev/peps/pep-0440/) versions
        (N[.N]+[{a|b|c|rc}N][.postN][.devN])
    * PerlVersionScheme which supports 2 numerical part versions where the second part is at least two digits
        (A.BB where A and B are integers and B is zero padded on the left.  For example:  1.02, 1.34, 1.567)

If you don't specify which version scheme the version instance uses, then it will use the first scheme from the
SupportedVersionSchemes list that successfully parses the version string

By default, Pep440VersionScheme is the supported scheme.  To change to a different list of schemes, use the
**Version.set_supported_version_schemes(schemes)**.  For example::

    from versio.version_scheme import Simple3VersionScheme, PerlVersionScheme
    Version.set_supported_version_schemes([Simple3VersionScheme, PerlVersionScheme])

In addition, you may define your own version scheme by extending VersionScheme.
"""

__docformat__ = 'restructuredtext en'

import re
from versio.comparable_mixin import ComparableMixin
from versio.version_scheme import Pep440VersionScheme


__all__ = ['Version']


class Version(ComparableMixin):
    """
    A version class that supports multiple versioning schemes, version comparisons,
    and version bumping.
    """

    # Version uses the SupportedVersionSchemes list when trying to parse a string where
    # the given scheme is None.  The parsing attempts will be sequentially thru this list
    # until a match is found.
    supported_version_schemes = [
        # Simple3VersionScheme,
        # Simple4VersionScheme,
        Pep440VersionScheme,
    ]

    def _cmpkey(self):
        """
        A key for comparisons required by ComparableMixin

        Here we just use the list of version parts.
        """
        return self.parts

    @classmethod
    def set_supported_version_schemes(cls, schemes):
        """
        Set the list of version schemes used when parsing a string.

        :param schemes:  list of version schemes.
        """
        cls.supported_version_schemes = list(schemes)

    def __init__(self, version_str=None, scheme=None):
        """
        Creates a version instance which is bound to a version scheme.

        :param version_str: the initial version as a string
        :type version_str: str or None
        :param scheme: the version scheme to use to parse the version string or None to try all supported
        version schemes.
        :type scheme: VersionScheme or None
        """
        self.scheme, self.parts = self._parse(version_str, scheme)
        if not self.scheme:
            raise AttributeError("Can not find supported scheme for \"{ver}\"".format(ver=version_str))
        if not self.parts:
            raise AttributeError("Can not parse \"{ver}\"".format(ver=version_str))

    def _parse(self, version_str, scheme):
        """
        Parse the given version string using the given version scheme.  If the given version scheme
        is None, then try all supported version schemes stopping with the first one able to successfully
        parse the version string.

        :param version_str: the version string to parse
        :type version_str: str
        :param scheme: the version scheme to use
        :type scheme: VersionScheme or None
        :return: the version scheme that can parse the version string, and the version parts as parsed.
        :rtype: VersionScheme or None, list
        """
        if scheme is None:
            for trial_scheme in self.supported_version_schemes:
                parts = self._parse_with_scheme(version_str, trial_scheme)
                if parts:
                    return trial_scheme, parts
        else:
            return scheme, self._parse_with_scheme(version_str, scheme)
        return None

    def _parse_with_scheme(self, version_str, scheme):
        """
        Parse the version string with the given version scheme.

        :param version_str: the version string to parse
        :type version_str: str
        :param scheme: the version scheme to use
        :type scheme: VersionScheme
        :returns the parts of the version identified with the regular expression or None.
        :rtype: list of str or None
        """
        if version_str is None:
            return False
        return scheme.parse(version_str)

    def __str__(self):
        """
        Render to version to a string.

        :return: the version as a string.
        :rtype: str
        """
        if self.parts:
            casts = self.scheme.format_types
            casts = casts + [str] * (len(self.scheme.fields) - len(casts))   # right fill with str types

            parts = [part or '' for part in self.parts]
            parts = parts + [''] * (len(self.scheme.fields) - len(parts))   # right fill with blanks

            def _type_cast(value, cast):
                """cast the given value to the given cast or str if cast is None"""
                if cast is None:
                    cast = str
                result = None
                try:
                    result = cast(value)
                except ValueError:
                    pass
                return result

            args = list(map(_type_cast, parts, casts))
            return self.scheme.format_str.format(*args)
        return "Unknown version"

    def bump(self, field_name=None, sub_index=-1):
        """
        Bump the given version field by 1.  If no field name is given,
        then bump the least significant field.

        :param field_name: the field name that matches one of the scheme's fields
        :type field_name: object
        :param sub_index: index in field
        :type sub_index: int
        :return: True on success
        :rtype: bool
        """
        if field_name is None:
            field_name = self.scheme.fields[-1]
        field_name = field_name.lower()

        # noinspection PyBroadException
        try:
            bumped = False
            index = self.scheme.fields.index(field_name)
            for idx, part in enumerate(self.parts):
                if idx == index:
                    bumped_part = self._bump_parse(field_name, part, sub_index)
                    if self.parts[idx] != bumped_part:
                        self.parts[idx] = bumped_part
                        bumped = True

                if idx > index:
                    self.parts[idx] = self.scheme.clear_value
            return bumped
        except (IndexError, ValueError):
            # not if fields, try subfields
            if field_name in self.scheme.subfields:
                return self.bump(*self.scheme.subfields[field_name])
            return False

    def _increment(self, field_name, value):
        """
        Increment the value for the given field.

        :param field_name: the field we are incrementing
        :type field_name: str
        :param value: the field's value
        :type value: int or str
        :return: the value after incrementing
        :rtype: int or str
        """
        if isinstance(value, int):
            value += 1
        if isinstance(value, str):
            if field_name in self.scheme.sequences:
                seq_list = self.scheme.sequences[field_name]
                if not value:
                    return seq_list[0]
                if value not in seq_list:
                    raise AttributeError('Can not bump version, the current value (%s) not in sequence constraints' %
                                         value)
                idx = seq_list.index(value) + 1
                if idx < len(seq_list):
                    return seq_list[idx]
                else:
                    raise IndexError('Can not increment past end of sequence')
            else:
                value = chr(ord(value) + 1)
        return value

    def _part_increment(self, field_name, sub_index, separator, sub_parts, clear_value):
        """
        Increment a version part, including handing parts to the right of the field being incremented.

        :param field_name: the field we are incrementing
        :type field_name: str
        :param sub_index: the index of the sub part we are incrementing
        :type sub_index: int
        :param separator: the separator between sub parts
        :type separator: str or None
        :param sub_parts: the sub parts of a version part
        :type sub_parts: list of int or str
        :param clear_value: the value to set parts to the right of this part to after incrementing.
        :type clear_value: str or None
        :return:
        :rtype:
        """
        sub_parts[sub_index] = self._increment(field_name, sub_parts[sub_index])
        if sub_index >= 0:
            for sub_idx in range(sub_index + 1, len(sub_parts)):
                sub_parts[sub_idx] = clear_value
        return separator.join([str(n) for n in sub_parts])

    def _bump_parse(self, field_name, part, sub_index):
        """
        Bump (increment) the given field of a version.

        :param field_name: the field we are incrementing
        :type field_name: str
        :param part: the version part being incremented
        :type part: str or None
        :param sub_index:
        :type sub_index:
        :return: the version part after incrementing
        :rtype: int or str or None
        """
        if part is None:
            value = self.scheme.clear_value or '1'
            return '{seq}{value}'.format(seq=self.scheme.sequences[field_name][0], value=value)

        match = re.match(r'^\d[\.\d]*(?<=\d)$', part)
        if match:
            # dotted numeric (ex: '1.2.3')
            return self._part_increment(field_name, sub_index, '.', [int(n) for n in part.split('.')],
                                        self.scheme.clear_value or '0')

        match = re.match(r'(\.?[a-zA-Z]*)(\d+)', part)
        if match:
            #  alpha + numeric (ex: 'a1', 'rc2', '.post3')
            return self._part_increment(field_name, sub_index, '', [match.group(1) or '', int(match.group(2))],
                                        self.scheme.clear_value or '1')

        return part
